pkg/deploy/registry/deploylog/rest.go

```diff
@@ -0,0 +1,181 @@
+	return c.WatchRcFn(ctx, label, field, rv)
+}
+
+type podGetter struct {
+	podsNamespacer kclient.PodsNamespacer
+}
+
+// Get is responsible for retrieving the deployer pod
+func (g *podGetter) Get(ctx kapi.Context, name string) (runtime.Object, error) {
+	ns, ok := kapi.NamespaceFrom(ctx)
+	if !ok {
+		return nil, errors.NewBadRequest("namespace parameter required.")
+	}
+	return g.podsNamespacer.Pods(ns).Get(name)
+}
+
+const defaultTimeout time.Duration = 10 * time.Second
+
+// NewREST creates a new REST for DeployLogs. It accepts a DeploymentClient and a pod
+// client to get necessary attributes to assemble the URL to which the request shall be
+// redirected in order to get the deployment logs.
+func NewREST(d DeploymentClient, pn kclient.PodsNamespacer, connectionInfo kclient.ConnectionInfoGetter) *REST {
+	return &REST{
+		Deploy:         d,
+		PodGetter:      &podGetter{pn},
+		ConnectionInfo: connectionInfo,
+		Timeout:        defaultTimeout,
+	}
+}
+
+// REST implements GetterWithOptions
+var _ = rest.GetterWithOptions(&REST{})
+
+// Get returns a streamer resource with the contents of the deployment log
+func (r *REST) Get(ctx kapi.Context, name string, opts runtime.Object) (runtime.Object, error) {
+	deployLogOpts, ok := opts.(*deployapi.DeploymentLogOptions)
+	if !ok {
+		return nil, errors.NewBadRequest("did not get an expected options.")
+	}
+
+	config, err := r.Deploy.GetDeploymentConfig(ctx, name)
+	if err != nil {
+		return nil, errors.NewNotFound("deploymentConfig", name)
+	}
+
+	deployment, err := r.Deploy.GetDeployment(ctx, deployutil.LatestDeploymentNameForConfig(config))
+	if err != nil {
+		// TODO: Fallback to latest-1?
+		return nil, errors.NewNotFound("deployment", deployutil.LatestDeploymentNameForConfig(config))
+	}
+
+	status := deployutil.DeploymentStatusFor(deployment)
+	switch status {
+	case deployapi.DeploymentStatusNew, deployapi.DeploymentStatusPending:
+		if deployLogOpts.NoWait {
+			glog.V(4).Infof("Deployment %s/%s is in %s state. No logs to retrieve yet.", deployment.Namespace, deployment.Name, status)
+			return &genericrest.LocationStreamer{}, nil
+		}
+		glog.V(4).Infof("Deployment %s/%s is in %s state, waiting for it to start...", deployment.Namespace, deployment.Name, status)
+		if err := r.waitForDeployment(ctx, deployment); err != nil {
+			return nil, err
+		}
+	}
+
+	deployPodName := deployutil.DeployerPodNameForDeployment(deployment.Name)
+	logOpts := &kapi.PodLogOptions{
+		Follow: deployLogOpts.Follow,
+	}
+	location, transport, err := pod.LogLocation(r.PodGetter, r.ConnectionInfo, ctx, deployPodName, logOpts)
+	if err != nil {
+		return nil, errors.NewBadRequest(err.Error())
+	}
+	return &genericrest.LocationStreamer{
+		Location:    location,
+		Transport:   transport,
+		ContentType: "text/plain",
+		Flush:       deployLogOpts.Follow,
+	}, nil
+}
+
+func (r *REST) waitForDeployment(ctx kapi.Context, deployment *kapi.ReplicationController) error {
```

Comment:
We've been down this path at least a couple times now (blocking while awaiting for a pod to enter some phase). The last time we did it was here:

https://github.com/openshift/origin/blob/master/pkg/deploy/strategy/support/lifecycle.go#L219

Using the `Watch` interface directly is hard to get right: although what you have here will work, it's not handling watch timeouts or interruptions or missed events. The `client/cache` package has all the structures to do this work for us. It's not semantically ideal for this case, but it definitely works well. What I did in the lifecycle code was to set up a `Reflector` populating a `Store` with the pod (or pods) to watch. Then your code can use `wait.Poll` on the `Store` to check the pod phase every once in a while. You still get server side watch efficiently at the expense of client side polling, but you no longer have to handle any of the fault tolerance yourself. I think it's worth it. We might be at a point where this specific use case could be extracted into a library. I know volumes code is doing something similar. Builds might be too but I haven't dug around yet.

An added benefit is your tests don't need to deal with setting up fake watches: they can manipulate a local `cache.Store` instance in the test and the `wait.Poll` calls will use it without any fuss.